stl/inc/deque

```diff
@@ -986,19 +986,58 @@ public:
     }
 
     void shrink_to_fit() {
-        size_type _Oldcapacity = _Block_size * _Mapsize();
-        size_type _Newcapacity = _Oldcapacity / 2;
+        if (empty()) {
+            if (_Map() != nullptr) {
+                _Reset_map();
+            }
+            return;
+        }
+
+        const auto _First_block_ptr = _Map() + static_cast<_Map_difference_type>(_Myoff() / _Block_size);
+        const auto _Last_block_ptr =
+            _Map() + static_cast<_Map_difference_type>((_Myoff() + _Mysize() - 1) / _Block_size + 1);
+        const auto _Used_block_count = static_cast<size_type>(_Last_block_ptr - _First_block_ptr);

```

Comment:
`_Last_block_ptr` is able to come before `_First_block_ptr`, as masking is needed. Try this:
```
        std::deque<int> deq(128);
        for (int i = 0; i < 120; ++i) {
            deq.pop_back();
        }
        for (int i = 0; i < 10; ++i) {
            deq.push_front(1);
        }
        deq.shrink_to_fit();
```

Reply:
Thanks! I totally ignored that fact that the internal map is a circular buffer.